### ISSUE TITLE
Ghost updates with flatpak remote-ls --updates

### ISSUE THREAD

**Issue by fabricemarie** [flatpak/flatpak#3748]
## Linux distribution and version

Fedora SilverBlue 32

## Flatpak version

1.6.4 and reproduced with 1.8.1

## Description of the problem

Listing available updates give a list of packages that need updates **and** packages that have already been fully updated:

```
# flatpak remote-ls --updates
Name                   Application ID                   Version       Branch       Arch         Origin
Code - OSS             com.visualstudio.code-oss        1.47.1        stable       x86_64       flathub
Baobab                 org.gnome.Baobab                               stable       x86_64       fedora
GNOME Calculator       org.gnome.Calculator             3.36.0        stable       x86_64       fedora
Calendar               org.gnome.Calendar               3.36.1        stable       x86_64       fedora
GNOME Characters       org.gnome.Characters                           stable       x86_64       fedora
Contacts               org.gnome.Contacts               3.36.1        stable       x86_64       fedora
Evince                 org.gnome.Evince                 3.36.3        stable       x86_64       fedora
File Roller            org.gnome.FileRoller                           stable       x86_64       fedora
GNOME Logs             org.gnome.Logs                   3.34.0        stable       x86_64       fedora
GNOME Maps             org.gnome.Maps                   3.36.2        stable       x86_64       fedora
GNOME Screenshot       org.gnome.Screenshot             3.36.0        stable       x86_64       fedora
Weather                org.gnome.Weather                3.36.1        stable       x86_64       fedora
GNOME Clocks           org.gnome.clocks                 3.36.2        stable       x86_64       fedora
Eye of GNOME           org.gnome.eog                    3.36.2        stable       x86_64       fedora
GNOME Fonts            org.gnome.font-viewer            3.32.0        stable       x86_64       fedora
gedit                  org.gnome.gedit                  3.36.2        stable       x86_64       fedora
Platform               org.fedoraproject.Platform                     f32          x86_64       fedora
```

when the actual list of updates (gathered with `flatpak update`) is:

```
# flatpak update
Looking for updates…

New org.gnome.clocks permissions:
    fallback-x11



        ID                                  Branch          Op          Remote           Download
 1.     org.gnome.font-viewer               stable          u           fedora           < 253.6 kB
 2.     org.gnome.eog                       stable          u           fedora             < 1.4 MB
 3.     org.gnome.clocks                    stable          u           fedora             < 1.1 MB
 4.     org.gnome.Weather                   stable          u           fedora             < 5.8 MB
 5.     org.gnome.FileRoller                stable          u           fedora           < 335.2 kB
 6.     org.gnome.Evince                    stable          u           fedora            < 17.0 MB
 7.     org.gnome.Contacts                  stable          u           fedora             < 6.0 MB
 8.     org.gnome.Characters                stable          u           fedora           < 347.0 kB
 9.     org.gnome.Calendar                  stable          u           fedora             < 5.9 MB
10.     com.visualstudio.code-oss           stable          u           flathub           < 85.8 MB

Proceed with these changes to the system installation? [Y/n]: n
```

## Steps to reproduce
Compare the output of `flatpak remote-ls --updates` with `flatpak update`

## Question

What is the recommended way to list updates without installing them?

**Comment by mwleeds** [comment]
>What is the recommended way to list updates without installing them?

You can run `flatpak update` and answer `no` at the `Proceed...` prompt, as you did.

That said, it looks like `remote-ls --updates` is misbehaving.

**Comment by mwleeds** [comment]
Could you post the output of `flatpak info org.gnome.Baobab`?

**Comment by mwleeds** [comment]
Looking at the code the only obvious reason `remote-ls --updates` would differ from `update` is if the set of installed subpaths is incomplete, but that would mostly apply to locale extensions so doesn't seem like the explanation here.

**Comment by fabricemarie** [comment]
> Could you post the output of `flatpak info org.gnome.Baobab`?

```
# flatpak info org.gnome.Baobab
          ID: org.gnome.Baobab
         Ref: app/org.gnome.Baobab/x86_64/stable
        Arch: x86_64
      Branch: stable
      Origin: fedora
  Collection: 
Installation: system
   Installed: 328.2 kB
     Runtime: org.fedoraproject.Platform/x86_64/f32
         Sdk: org.fedoraproject.Sdk/x86_64/f32

      Commit: 5cceef41fb525e21210f3c08eec83f2794523ee59a5a03f852a5c628d059b0cf
     Subject: Export org.gnome.Baobab
        Date: 2020-03-31 17:52:51 +0000
      Alt-id: 2433aee27a9fdeaf7292ab2d072b22729fa0add69c0fff40d34c616b1db3f5e1
```

And for example with Map which was updated as well and still shows as a ghost:
```
# flatpak info org.gnome.Maps

GNOME Maps - Find places around the world

          ID: org.gnome.Maps
         Ref: app/org.gnome.Maps/x86_64/stable
        Arch: x86_64
      Branch: stable
     Version: 3.36.2
     License: GPL-2.0-or-later
      Origin: fedora
  Collection: 
Installation: system
   Installed: 20.6 MB
     Runtime: org.fedoraproject.Platform/x86_64/f32
         Sdk: org.fedoraproject.Sdk/x86_64/f32

      Commit: 109bdae272aff90ded5db0b09faac7b05edcda92e7c6fc4490c348d40e47a5ad
     Subject: Export org.gnome.Maps
        Date: 2020-06-04 12:54:36 +0000
      Alt-id: 540ff90441f12390cdc1acab39ec00fe64e62870f9e38f1b377a0b77baf11e7d

```

**Comment by fabricemarie** [comment]
Flatpak now simply refuse to update as-is:

```
# flatpak update --appstream
Updating appstream data for remote fedora
Updating appstream data for remote flathub

Not sure if that helps.

# flatpak update -v
F: No installations directory in /etc/flatpak/installations.d. Skipping
F: Opening system flatpak installation at path /var/lib/flatpak
F: Opening user flatpak installation at path /root/.local/share/flatpak
Looking for updates…
F: Fetching summary file for remote ‘flathub’
F: Transaction: update flathub:app/com.visualstudio.code-oss/x86_64/stable[$old]
F: Transaction: update flathub:app/org.gimp.GIMP/x86_64/stable[$old]
F: flatpak_http_should_retry_request: error: 359:0 Reusing cached value, n_retries_remaining: 5
F: Transaction: update fedora:app/org.gnome.Baobab/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.Calculator/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.Calendar/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.Characters/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.Contacts/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.Evince/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.FileRoller/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.Logs/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.Maps/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.Screenshot/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.Weather/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.clocks/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.eog/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.font-viewer/x86_64/stable[$old]
F: Transaction: update fedora:app/org.gnome.gedit/x86_64/stable[$old]
F: Transaction: update fedora:runtime/org.fedoraproject.Platform/x86_64/f32[$old]
F: Transaction: update flathub:runtime/org.freedesktop.Platform.GL.default/x86_64/19.08[$old]
F: Transaction: update flathub:runtime/org.freedesktop.Platform.openh264/x86_64/2.0[$old]
F: Transaction: update flathub:runtime/org.freedesktop.Sdk.Locale/x86_64/19.08[/en]
F: Transaction: update flathub:runtime/org.freedesktop.Sdk/x86_64/19.08[$old]
F: Transaction: update flathub:runtime/org.gimp.GIMP.Manual/x86_64/2.10[$old]
F: Transaction: update flathub:runtime/org.gnome.Platform.Locale/x86_64/3.36[/en]
F: Transaction: update flathub:runtime/org.gnome.Platform/x86_64/3.36[$old]
F: Looking for remote metadata updates for flathub
F: Looking for remote metadata updates for fedora
F: marking op update:runtime/org.gnome.Platform/x86_64/3.36 resolved to 6ce207edb05a8c345b5cf90263fd1f3d931a82c754645599d819a0349797d412
F: marking op update:runtime/org.gnome.Platform.Locale/x86_64/3.36 resolved to ecc783db2eaafd0a8d89f2dd409e179f23ab7fc33418550c7450bbe44ea22be3
F: marking op update:runtime/org.gimp.GIMP.Manual/x86_64/2.10 resolved to 1a0a3b85b68d482998cf8044ddfc56dca74916fb9471257d94bc4ff6f8795800
F: marking op update:runtime/org.freedesktop.Sdk/x86_64/19.08 resolved to ee4ecf9fbcf2de32d2c500e3c65e4460ada89ab91217e76a122c431cec63e54c
F: marking op update:runtime/org.freedesktop.Sdk.Locale/x86_64/19.08 resolved to 396bcbf59922e014e85aa2ca8ae2dde7a94cd575642c7878debd1c076b0ea9f2
F: marking op update:runtime/org.freedesktop.Platform.openh264/x86_64/2.0 resolved to 73f998362a6fc0d57e0c7e83e928d32b0ec14d10d0d94291033976bdcecc6b6b
F: marking op update:runtime/org.freedesktop.Platform.GL.default/x86_64/19.08 resolved to f68c8e930d048d0490c605d992a82e11499c78cf374a67ef69fa15a4210b636b
F: marking op update:runtime/org.fedoraproject.Platform/x86_64/f32 resolved to fafc9e1f23d78abf3b0780a8fa28dbf1ec6cca07fec7d1042490861796cad22e
F: marking op update:app/org.gnome.gedit/x86_64/stable resolved to 8a213d47b89699b68b127a613ff17b674d427dd694e4715f3d842a5ac21588af
F: marking op update:app/org.gnome.font-viewer/x86_64/stable resolved to 0723530f7aa735a84cc5b3915c90b752f6cdb66e86562b0e6d8a452ea313a498
F: marking op update:app/org.gnome.eog/x86_64/stable resolved to 74ca6382f7cc8ec24cc1fe4a37b56cf43a4f2724d6aca0a413a1c77d9da8de07
F: marking op update:app/org.gnome.clocks/x86_64/stable resolved to b9d1cf5b7e662b2a4a973de5285345470c733dcbefc82aa105939c8846b53ba7
F: marking op update:app/org.gnome.Weather/x86_64/stable resolved to 9f2b073b1869df5aab58a8696f0d59bbdf53e377e3d838aa33d3aea04346444d
F: marking op update:app/org.gnome.Screenshot/x86_64/stable resolved to 68c5db3ddba3270994b2378a47a71fcf1bced1b0e3cfc0cdc81ebc2642e294ad
F: marking op update:app/org.gnome.Maps/x86_64/stable resolved to 540ff90441f12390cdc1acab39ec00fe64e62870f9e38f1b377a0b77baf11e7d
F: marking op update:app/org.gnome.Logs/x86_64/stable resolved to 0601ae729b953aed688a40092937a8d9ac218b67f0c9c491021e6c80b3556690
F: marking op update:app/org.gnome.FileRoller/x86_64/stable resolved to 60a42ba3d38a761a5a99857176b08d95e141dc968b90dc2d25cebfc0cb3b7be5
F: marking op update:app/org.gnome.Evince/x86_64/stable resolved to 77c5bc2973d26dc51ce261cf6354fa0802876e3c9c4f2d7bbc97ccc55bcabb00
F: marking op update:app/org.gnome.Contacts/x86_64/stable resolved to 7876045e2e6d20f5412e882efb966a27dbee411ad49b3749d34753f3243d8ece
F: marking op update:app/org.gnome.Characters/x86_64/stable resolved to f6f6e12610b8b863bbb9e09cbf4f2474fbd77f809d6066b1f632e592d6df3325
F: marking op update:app/org.gnome.Calendar/x86_64/stable resolved to a10c2824858559282a638f4d445c258a0845b601c9d7d05d4c2d7c82ddea21f7
F: marking op update:app/org.gnome.Calculator/x86_64/stable resolved to 797dc8b908de6039eead1e03c74b8bb19413f07e96201baa5822cd53f7a9fc9d
F: marking op update:app/org.gnome.Baobab/x86_64/stable resolved to 2433aee27a9fdeaf7292ab2d072b22729fa0add69c0fff40d34c616b1db3f5e1
F: marking op update:app/org.gimp.GIMP/x86_64/stable resolved to a69a3d31490bb01c325cc19c58f3cdc1029c74476a13bf5a7696e7cbd92a7fd0
F: marking op update:app/com.visualstudio.code-oss/x86_64/stable resolved to efda261a29d08a6c887d66d360f4840b57327205d67634a6ae4f18d71faa973f
F: Transaction: install/update flathub:runtime/org.gnome.Platform.Locale/x86_64/3.36[/en]
F: Transaction: install/update flathub:runtime/org.freedesktop.Platform.GL.default/x86_64/19.08[*]
F: Transaction: install/update flathub:runtime/org.freedesktop.Platform.openh264/x86_64/2.0[*]
F: Transaction: install/update flathub:runtime/org.freedesktop.Platform.GL.default/x86_64/19.08[*]
F: Transaction: install/update flathub:runtime/org.freedesktop.Sdk.Locale/x86_64/19.08[/en]
F: Transaction: install/update flathub:runtime/org.freedesktop.Platform.openh264/x86_64/2.0[*]
F: Transaction: install/update flathub:runtime/org.gimp.GIMP.Manual/x86_64/2.10[/en]
F: Requesting tokens for remote fedora: (app/org.gnome.font-viewer/x86_64/stable, 0723530f7aa735a84cc5b3915c90b752f6cdb66e86562b0e6d8a452ea313a498 1), (app/org.gnome.eog/x86_64/stable, 74ca6382f7cc8ec24cc1fe4a37b56cf43a4f2724d6aca0a413a1c77d9da8de07 1), (app/org.gnome.clocks/x86_64/stable, b9d1cf5b7e662b2a4a973de5285345470c733dcbefc82aa105939c8846b53ba7 1), (app/org.gnome.Weather/x86_64/stable, 9f2b073b1869df5aab58a8696f0d59bbdf53e377e3d838aa33d3aea04346444d 1), (app/org.gnome.FileRoller/x86_64/stable, 60a42ba3d38a761a5a99857176b08d95e141dc968b90dc2d25cebfc0cb3b7be5 1), (app/org.gnome.Evince/x86_64/stable, 77c5bc2973d26dc51ce261cf6354fa0802876e3c9c4f2d7bbc97ccc55bcabb00 1), (app/org.gnome.Contacts/x86_64/stable, 7876045e2e6d20f5412e882efb966a27dbee411ad49b3749d34753f3243d8ece 1), (app/org.gnome.Characters/x86_64/stable, f6f6e12610b8b863bbb9e09cbf4f2474fbd77f809d6066b1f632e592d6df3325 1), (app/org.gnome.Calendar/x86_64/stable, a10c2824858559282a638f4d445c258a0845b601c9d7d05d4c2d7c82ddea21f7 1)
F: Response from request_tokens: 2 - {'error-message': <'Unexpected response status 503 from repo'>, 'error-code': <-1>}

error: Failed to get tokens for ref: Unexpected response status 503 from repo
```

**Comment by mwleeds** [comment]
>``` 
>error: Failed to get tokens for ref: Unexpected response status 503 from repo
> ```

Looks like the same error message as in #3753

**Comment by fabricemarie** [comment]
Now with `Flatpak 1.8.2` the issue becomes clearer:

```Shell
# flatpak remote-ls --updates
Name                                                  Application ID                   Version                   Branch          Arch
Firefox                                               org.mozilla.firefox              81.0.1                    stable          x86_64
Zoom                                                  us.zoom.Zoom                     5.3.469451.0927           stable          x86_64
GNOME Application Platform version 3.36               org.gnome.Platform                                         3.36            x86_64
GNOME Software Development Kit version 3.36           org.gnome.Sdk                                              3.36            x86_64
Adwaita theme                                         org.kde.KStyle.Adwaita                                     5.14            x86_64
Adwaita theme                                         org.kde.KStyle.Adwaita                                     5.15            x86_64
KDE Application Platform                              org.kde.Platform                                           5.14            x86_64
KDE Application Platform                              org.kde.Platform                                           5.15            x86_64
```

whereas `flatpak update` shows:
```Shell
# flatpak update
Looking for updates…


        ID                                   Branch          Op          Remote           Download
 1.     us.zoom.Zoom                         stable          u           flathub           < 40.3 MB
 2.     org.gnome.Platform.Locale            3.36            u           flathub          < 323.1 MB (partial)
 3.     org.gnome.Platform                   3.36            u           flathub          < 326.0 MB
 4.     org.gnome.Sdk.Locale                 3.36            u           flathub          < 327.2 MB (partial)
 5.     org.gnome.Sdk                        3.36            u           flathub          < 673.1 MB
 6.     org.kde.KStyle.Adwaita               5.14            u           flathub            < 6.6 MB
 7.     org.kde.KStyle.Adwaita               5.15            u           flathub            < 6.1 MB
 8.     org.kde.Platform.Locale              5.14            u           flathub          < 337.6 MB (partial)
 9.     org.kde.Platform                     5.14            u           flathub          < 362.5 MB
10.     org.kde.Platform.Locale              5.15            u           flathub          < 341.1 MB (partial)
11.     org.kde.Platform                     5.15            u           flathub          < 388.3 MB
12.     org.mozilla.firefox.Locale           stable          u           flathub           < 44.1 MB (partial)
13.     org.mozilla.firefox                  stable          u           flathub           < 79.9 MB
```

The 5 updates not listed by `flatpak remote-ls --updates` are the updates marked as `(partial)`.

What are these partial updates and is it normal that they aren't shown when running `flatpak remote-ls --updates`?
Is there another way to list all updates without installing them? (something like `flatpak update --no` to answer no and only get the output of the command)

**Comment by fabricemarie** [comment]
`flatpak remote-ls --updates --all` show the partial ones too actually. Missed it in the man page.

The issue of ghost updates I originally highlighted seems fixed with `flatpak 1.8.2` and I no longer have inconsistent lists between `flatpak update` and `flatpak remote-ls --updates` after I manually install updates. Thanks a lot.
